Fix #3124 - Move enums to top level schema definitions

mtraynham opened the pull request:
This fixes #3124 by moving enums to top level schema properties, original inspiration from #2996.

@frantuma I know the tests are likely going to break because this changes how enums work.  I was hoping to get your feedback on how to make this configurable, like #2996 originally proposed.  Should it be an annotation on the class or a configuration property of the OpenAPI?

frantuma commented:
@mtraynham thanks for looking into this! I share here some thoughts.

1. I guess the changed code is not sufficient (I also don't understand the change to `Components`) to address the issue; we would need to convert the resolved enum into a ref and add that to the components sections, as done for non primitive types (depending on scenario); also specific tests would need to be added

2. About the way to control behaviour, I can see 2 ways, independent from each other:

A. implement a new annotation e.g `@EnumAsRef`(name = "")` and check it in code for point 1 above.

B. add it as a `Configuration` member; this is more complicated, as atm configuration is not passed down to `ModelConverter` instances (e.g. `ModelResolver`) and related classes; we would probably need to slightly refactor converter related code to gain access to config property; the alternative based on shared static variables sounds more limited and not really consistent, also not applicable e.g. in maven plugin or in scenarios with no java code related to swagger-core config.

mtraynham commented:
@frantuma heh, I also do not entirely understand why the change in #2996 to `Components` is necessary, but it does seem to be.  I removed it and Swagger instead produces an empty block for enums with no properties.

I'm not against the `@EnumAsRef` or maybe represented as `@Schema(enumAsRef=true)`, my only concern there is consistency throughout a user's API.  But, I also am unfamiliar with how this code works, so it's likely the safer route for me than configuration changes.

xiaoxiangmoe commented:
@mtraynham What is the current state of this feature? Thanks.

mtraynham commented:
@xiaoxiangmoe This works, but I haven't really heard back from @frantuma.

mtraynham commented:
Alright, I think the initial implementation was a bit strange, since it had a `Components` change which nulled the `$ref`.  But after following some of this recursion in `ModelResolver`, I think my latest push is a better implementation.  It simply adds the enum model as a top level schema to the context and returns a simple `Schema` with a `ref`.  I reverted the `Components` change and added a test for this.

mtraynham commented:
By the way, I have found that if you have an enum that uses a different method of serializing and deserializing with say Jackson, such as an `integer` enum, you can accomplish top level enums in a similar fashion with:

```java
    @Schema(enumAsRef = true, type = "integer", allowableValues = {"1", "2"})
    public enum NumberedEnum {
        NUMBER_1(1),
        NUMBER_2(2);

        private final Integer value;
        NumberedEnum(Integer value) {
            this.value = value;
        }

        @JsonCreator(mode = JsonCreator.Mode.DELEGATING)
        public static NumberedEnum fromValue(Integer value) {
            for (NumberedEnum other : NumberedEnum.values()) {
                if (other.value.equals(value)) {
                    return other;
                }
            }
            throw new IllegalArgumentException("Unsupported value " + value);
        }
    }
```

xiaoxiangmoe commented:
Is there any feedback on this pr? @frantuma

xiaoxiangmoe commented:
@frantuma Please have a look. Thanks.

mtraynham commented:
@frantuma any updates on this?

frantuma commented:
It's in our radar, we plan to have it in for next release

xiaoxiangmoe commented:
@frantuma When will the next edition be released?

frantuma commented:
No exact ETA, but within 2 weeks

xiaoxiangmoe commented:
@frantuma any updates on this?

frantuma commented:
Thanks a lot! this is included and replaced by #3352 which adds also global support via sys prop / static var
